### ISSUE TITLE
Model Distillation

### ISSUE THREAD

**Issue by lunixbochs** [facebookresearch/wav2letter#289]
I'm working on a distillation pass to piggyback on a larger model to generate a small high quality model, since that's going to be easier than quantization through flashlight for now.

Can a dev validate my approach?

1. Copy `Train` to `Distill`
2. Add FLAGS_am to load an already-trained "big" model.
3. Don't create or load a criterion/critopt for the small model. When saving the small model, just save the big trained model's criterion/critopt.
4. When training:
  a. Don't train the criterion.
  b. small_emissions = small->forward(input)
  c. big_emissions = big->forward(input)
  d. loss = cross-entropy diff of emissions? I don't actually know how to do this part yet
  e. netopt->zeroGrad(); loss.backward(); netopt->step();

Do these steps look correct?
Any advice on calculating and creating the loss object within Wav2letter?

**Comment by lunixbochs** [comment]
Ok, I have this now, but I'm not sure if it's correct (or even if I understand cross-entropy loss):
```
// forward
auto output = ntwrk->forward({fl::input(sample[kInputIdx])}).front();
auto refOutput = refNtwrk->forward({fl::input(sample[kInputIdx])}).front();
af::sync();
// cross-entropy loss
auto loss = fl::sum((output * -1) * fl::log(refOutput), {0});
af::sync();
```
Just going to try it :)

**Comment by tlikhomanenko** [comment]
Hi @lunixbochs,

Your steps looks good. 

For loss calculation your can try to use existing losses in flashlight https://fl.readthedocs.io/en/latest/modules.html#losses or define in a similar way your own.

Be care what exactly emissions are: scores, probabilities or log probabilities.
In case of probabilities you can use loss you defined above (switch the probs of reference model and small model:`loss = -fl::sum(refOutput * fl::log(output))`. 

During distillation it is better to approximate scores (before softmax) because in this case you don't loose information by mapping into probs in the interval (0, 1). In this case you minimize MSE loss. (A good explanation and info you can find, for example, in this paper https://arxiv.org/pdf/1503.02531.pdf)

**Comment by lunixbochs** [comment]
I didn't see anywhere the librispeech recipe would be doing softmax, so I tried the MSE loss function in both orders:
`fl::MeanSquaredError MSE;`
`auto loss = MSE.forward(output, refOutput);`
and `auto loss = MSE.forward(refOutput, output);`

I ran this on a very good (~2.6%) pretrained large model, and an ok (15%) pretrained small model, and the resulting model just outputs E and spaces, so I'm missing something still

**Comment by tlikhomanenko** [comment]
> I didn't see anywhere the librispeech recipe would be doing softmax, so I tried the MSE loss function in both orders:
fl::MeanSquaredError MSE;
auto loss = MSE.forward(output, refOutput);
and auto loss = MSE.forward(refOutput, output);

If you are using standard recipes, then yes, the softmax is done inside the criterion calculation.
MSE doesn't depend on the order of arguments (not true for cross-entropy). So, both your calls will produce the same thing.

> I ran this on a very good (~2.6%) pretrained large model as the reference and an ok (15%) pretrained small model, and the resulting model just outputs E and spaces, so I'm missing something still

Did you try to run test.cpp on the trained small model (when you trained it on MSE loss)? What was the MSE loss between outputs of reference and small models?

> Edit: ok, I realize the weights aren't going to match without a softmax. I couldn't get softmax to make a non-NaN loss because of the log, so I'm trying to train from scratch and see if it converges now

Didn't get this.

**Comment by lunixbochs** [comment]
If I start training with MSE loss, I see a loss of around 30, which goes down to 15 pretty quickly. I thought something was horribly wrong because the TER is claimed around 600 during Training, but it looks like the trained model actually gets a TER of ~40 when run externally against Test after a couple epochs of Distill, so maybe it's working and I just need to leave it alone for longer until the loss drops more.

My Distill code is here: https://github.com/talonvoice/wav2letter/blob/master/Distill.cpp#L481-L488

**Comment by tlikhomanenko** [comment]
Could you send the log of training?

- refNetwork should be in eval mode (maybe you are propagating through it too).
- `auto refOutput = refNtwrk->forward({fl::input(sample[kInputIdx])}).front();` -> `auto refOutput = refNtwrk->forward({fl::noGrad(sample[kInputIdx])}).front();`

Also it seems, that loss is still pretty high.

**Comment by lunixbochs** [comment]
I changed refNtwrk to eval mode and ran again, here's the log for the distillation: https://bochs.info/p/ny6xhc

We're at epoch 127 because I pretrained the small network. The TERs listed are completely wrong for some reason, maybe I broke the invocation of criterion / viterbiPath? The actual TER when I run a Test is sub-100, but still kinda bad like 25%

I didn't wire up loss printing for the test datasets so those are zero.

I think Distilling between two separately-trained networks without a softmax to normalize their output weights relative to each other doesn't seem to work very well, or at least not very quickly.

Perhaps training a Distilled network from scratch will converge faster, and I should just fix the TER output so I can see that it's making reasonable progress.

**Comment by lunixbochs** [comment]
Ok... at epoch 300 it's still making progress, down to LER of 12% on a test-clean segment with the MSE loss down to ~7, and the small model's output is improving over time in obvious and positive ways. TER during training is still wrong but doesn't seem to affect anything.
